tlv/stream.go

```diff
@@ -239,14 +240,17 @@ func (s *Stream) Decode(r io.Reader) error {
 			// We'll convert any EOFs to ErrUnexpectedEOF, since this
 			// results in an invalid record.
 			case err == io.EOF:
-				return io.ErrUnexpectedEOF
+				return nil, io.ErrUnexpectedEOF
 
 			// Other unexpected errors.
 			case err != nil:
-				return err
+				return nil, err
 			}
 		}
 
+		// Record the successfully decoded or ignored type.
+		parsedTypes[typ] = struct{}{}
```

Comment:
I don't know how concerned we are with performance of this. An idea could be to let the caller pass in an empty `parsedTypes` map. Here we could skip the assignment if the map is `nil`.

Reply:
good idea, maybe it should be a separate method, e.g. `DecodeWithParsedTypes` that can call into the inner method and pass nil or non-nil map

Reply:
That is better indeed, to prevent mistakes with passing a reused map.